### PR TITLE
Refactor val cadence

### PR DESCRIPTION
**Pull request by anna-grim**

### CODE REVIEW

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 1]
```diff
@@ -147,7 +115,7 @@ def __init__(
         self.scaler = torch.amp.GradScaler("cuda", enabled=self.use_amp)
 
     # --- Core Routines ---
-    def run(self, train_dataset, val_dataset):
+    def __call__(self, train_dataset, val_dataset):
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: This breaks train_bm4dnet.py, see https://github.com/AllenNeuralDynamics/aind-exaspim-image-compression/blob/f5b35d4dcf68cec06c650ba1887c68b00b5977e0/scripts/train_bm4dnet.py#L109

If this is to be a callable it needs to be updated to `trainer(train_dataset, val_dataset)`

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 2]
```diff
@@ -75,7 +47,7 @@ def __init__(
         fg_weight=20.0,
         num_workers=None,
         prefetch=2,
-        val_every=1,
+        val_every=1000,
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py:51]: train_bm4dnet.py still uses `val_every=1`, meaning that validation will  be done on every gradient step, see https://github.com/AllenNeuralDynamics/aind-exaspim-image-compression/blob/f5b35d4dcf68cec06c650ba1887c68b00b5977e0/scripts/train_bm4dnet.py#L144

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 3]
```diff
@@ -190,75 +148,81 @@ def run(self, train_dataset, val_dataset):
...
-        Performs a single training epoch over the provided DataLoader.
-
-        Parameters
-        ----------
-        train_dataloader : torch.utils.data.DataLoader
-            DataLoader for the training dataset.
-        epoch : int
-            Current training epoch.
+        # Create learning rate scheduler
+        total_steps = self.max_epochs * len(train_dataset) // self.batch_size
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: The total steps are being floored by batch_size here, but the `DataLoader` instead uses the ceiling, see https://github.com/AllenNeuralDynamics/aind-exaspim-image-compression/blob/f5b35d4dcf68cec06c650ba1887c68b00b5977e0/src/aind_exaspim_image_compression/machine_learning/data_handling.py#L1252-L1255
This can cause the cosine LR to rise again at end of training when the dataset is not divisible (which is normally the case, e.g., 30K patches with batch 32). 
Change this to `total_steps = self.max_epochs * ceil(len(train_dataset) / self.batch_size)`

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 4]
```diff
@@ -190,75 +148,81 @@ def run(self, train_dataset, val_dataset):
...
-        self.best_score = np.inf
-        for epoch in range(self.max_epochs):
-            # Train
-            train_dataloader.set_epoch(epoch)
-            train_loss = self.train_step(train_dataloader, epoch)
-
-            # Validate every val_every epochs (and always on the final epoch);
-            # the count-space metrics are CPU-bound, so a large validation set
-            # is only cheap when it is not run every epoch.
-            is_last = epoch == self.max_epochs - 1
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py:203]: The `is_last` condition should not be removed. Since validation now occurs only on exact `val_every` step boundaries,  if a run ends before the first boundary or between boundaries, the final (and potentially best) model is never evaluated or saved

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 5]
```diff
@@ -190,75 +148,81 @@ def run(self, train_dataset, val_dataset):
...
+                running_steps += 1
+                step += 1
+
+                # Validate (if applicable)
+                if step % self.val_every == 0:
+                    val_loss, val_cratio, val_score = self.validate(
+                        val_dataloader, step,
+                    )
+                    is_best = val_score < best_score
+                    if epoch > 0 and is_best:
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: `epoch > 0` is now the wrong condition to check if we are validating every N steps. With a large cache pool, all of epoch 0's validations save nothing.

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 6]
```diff
@@ -190,75 +148,81 @@ def run(self, train_dataset, val_dataset):
...
+                train_loss = self.train_step(x, y, fg_mask)
+                scheduler.step()
+
+                running_loss += train_loss
+                running_steps += 1
+                step += 1
+
+                # Validate (if applicable)
+                if step % self.val_every == 0:
+                    val_loss, val_cratio, val_score = self.validate(
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: When the validation dataset is empty, validate() returns `False` as its `val_score`. which is now being interpreted as a numeric score. Remember that in Python, `False < 1 == True` and `False == 0` is `True`

Therefore, on a validation boundary after epoch 0 this marks the empty result as score 0, potentially reports it as best, and saves a misleading checkpoint instead of skipping checkpoint selection as before.

**Review by carshadi — CHANGES_REQUESTED** [review]

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 1 above)
Comment (carshadi) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: This also breaks the unit tests so I would just revert unless you feel strongly about it being a callable.

Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: Good to know, I changed it back

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 2 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py:51]: Changed this too

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 3 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: Great catch, i updated this too

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 4 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py:203]: I added a final model validation at the end of the training loop

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 5 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: The simplest solution here is to save after every model validation, maybe that's the right choice since we can ultimately choose how many ckpts we want to evaluate after training

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 6 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: I updated the validate routine to return None if the validation dataset is empty

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 7]
```diff
@@ -190,93 +149,118 @@ def run(self, train_dataset, val_dataset):
...
+                        val_dataloader, step,
+                    )
+                    self.save_model(step, val_score)
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: This will crash when `val_score` is `None`

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 8]
```diff
@@ -190,93 +149,118 @@ def run(self, train_dataset, val_dataset):
...
+
+                    running_loss = 0
+                    running_steps = 0
+                    self.model.train()
+
+            # Final model validation
+            val_loss, val_cratio, val_score = self.validate(
+                val_dataloader, step,
+            )
+            self.save_model(step, val_score)
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: Same issue here

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 9]
```diff
@@ -190,93 +149,118 @@ def run(self, train_dataset, val_dataset):
...
+                        f"val_cratio={val_cratio:.5f}, "
+                        f"val_score={val_score:.5f}"
+                        f"{suffix}"
+                    )
+
+                    running_loss = 0
+                    running_steps = 0
+                    self.model.train()
+
+            # Final model validation
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: The final validation is inside the epoch loop, so it runs after every epoch regardless of val_every. It can also validate and save twice when an epoch ends on a validation boundary. Move this block outside the outer loop

**Review by carshadi — CHANGES_REQUESTED** [review]
Only a couple issues remain, please see the inline comments. 

On the whole, I would make validation a single guarded operation, then call it periodically and once after training only when the final step was not already validated.

  The key changes are:

  1. Never pass None to save_model() or numeric formatting.
  2. Track the last validated step.
  3. Dedent final validation outside the epoch loop.
  4. Skip duplicate final validation when the last step was already a cadence boundary.

  Conceptually:
```python
  step = 0
  best_score = np.inf
  last_validated_step = None

  def validate_and_checkpoint(step, label):
      nonlocal best_score, last_validated_step

      val_loss, val_cratio, val_score = self.validate(
          val_dataloader, step
      )
      last_validated_step = step

      if val_score is None:
          print(f"{label}: validation skipped (empty validation dataset)")
          return False

      is_best = val_score < best_score
      best_score = min(best_score, val_score)
      self.save_model(step, val_score)

      print(
          f"{label}: "
          f"val_loss={val_loss:.5f}, "
          f"val_cratio={val_cratio:.5f}, "
          f"val_score={val_score:.5f}"
          f"{' - New Best!' if is_best else ''}"
      )
      return is_best
```
  Use it at periodic boundaries:
```python
  if step % self.val_every == 0:
      avg_train_loss = running_loss / running_steps
      self.writer.add_scalar("train_loss", avg_train_loss, step)

      validate_and_checkpoint(step, f"Step {step}")

      running_loss = 0
      running_steps = 0
      self.model.train()
```
  Then, crucially, place this after the entire outer epoch loop:
```python
  # Outside: for epoch in range(self.max_epochs)
  if last_validated_step != step:
      validate_and_checkpoint(step, "Final model validation")
```
  This produces exactly one final validation, avoids duplicate checkpoints at cadence boundaries, and cleanly skips checkpoint selection when validation is empty. 

It would also be nice if we can add regression tests for:
  - A run shorter than val_every still validates and saves once.
  - A run ending exactly on val_every does not validate twice.
  - Multiple epochs do not trigger unconditional epoch-end validation.
  - An empty validation dataset neither saves nor raises an exception.

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 7 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: updated save_model to check if val_score is None and defaults to np.inf if it is

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 8 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: covered by earlier comment

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 9 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: Fixed!

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 10]
```diff
@@ -43,7 +42,7 @@ def __init__(
         max_epochs=400,
         model=None,
         use_amp=True,
-        use_amp_validation=False,
+        use_amp_val=False,
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: Changing this argument name breaks train_bm4dnet.py, which still uses the old name https://github.com/AllenNeuralDynamics/aind-exaspim-image-compression/blob/085637b740fcf47340fb2da272e379db5a5e3bc5/scripts/train_bm4dnet.py#L85
Changing this doesn't seem relevant to the scope of this PR either.

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 11]
```diff
@@ -435,7 +423,7 @@ def save_config(self, config):
             "val_every": self.val_every,
             "seed": self.seed,
             "use_amp": self.use_amp,
-            "use_amp_validation": self.use_amp_validation,
+            "use_amp_val": self.use_amp_val,
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: See above comment

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 12]
```diff
@@ -458,6 +446,7 @@ def save_model(self, epoch, score):
             embedded in the filename so checkpoints can be ranked offline.
         """
         date = datetime.today().strftime("%Y%m%d")
+        score = score or np.inf
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: `score = score or np.inf` changes a legitimate score of 0.0 to infinity. Since None is already filtered by `validate_and_checkpoint`, this line should simply be removed.

File: src/aind_exaspim_image_compression/machine_learning/train.py [hunk 13]
```diff
@@ -150,76 +149,48 @@ def run(self, train_dataset, val_dataset):
         )
 
         # Create learning rate scheduler
-        total_steps = self.max_epochs * math.ceil(len(train_dataset) / self.batch_size)
-        num_validations = math.ceil(total_steps / self.val_every)
+        total_steps = self.max_epochs * len(train_dataloader)
+        expected_validations = total_steps // self.val_every
```
Comment (carshadi) [inline src/aind_exaspim_image_compression/machine_learning/train.py]: `expected_validations = total_steps // val_every` undercounts by one whenever final validation is required; use `ceil(total_steps / val_every)`

**Review by carshadi — CHANGES_REQUESTED** [review]
This is close. In general, we should try to avoid introducing breaking public API changes that are unrelated to the scope of a PR, and if so they should be tested.

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 10 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: Done

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 11 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: Done

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 12 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: That's a good catch, just changed it to the conditional

score = np.inf if score is None else score

File: src/aind_exaspim_image_compression/machine_learning/train.py (hunk 13 above)
Comment (anna-grim) [reply src/aind_exaspim_image_compression/machine_learning/train.py]: Done

**Review by carshadi — APPROVED** [review]
Thanks @anna-grim !
